File: frontend/src/services/encomiendaService.ts
import { apiRequest } from './authService';
import type { 
  Encomienda, 
  EncomiendaFilters, 
  CreateEncomiendaRequest, 
  UpdateEncomiendaRequest,
  EncomiendaStats,
  PaginatedResponse,
  ApiResponse,
  StripePaymentIntent,
  ConfirmPaymentRequest
} from '../types/encomienda';

// ✅ INTERFACE PARA LA RESPUESTA DEL BACKEND
interface BackendEncomienda {
  id: number;
  codigo_seguimiento: string;
  remitente_nombre: string;
  remitente_telefono: string;
  remitente_direccion?: string;
  destinatario_nombre: string;
  destinatario_telefono: string;
  destino_ciudad: string;
  destino_direccion: string;
  descripcion: string;
  peso: number | string;
  precio: number | string;
  estado: string;
  fecha_creacion: string;
  fecha_entrega_estimada?: string;
  fecha_entrega_real?: string;
  conductor_asignado?: number;
  conductor_nombre?: string;
  notas?: string;
  creado_por?: number;
  pago_info?: any;
  estado_pago?: string;
  metodo_pago?: string;
  seguimientos?: any[];
  pago?: number;
  puede_ser_asignada?: boolean;
  puede_ser_entregada?: boolean;
  creado_por_nombre?: string;
  conductor_info?: any;
  pago_detalle?: any;
}

// ✅ FUNCIÓN PARA CALCULAR PRECIO AUTOMÁTICAMENTE
const calcularPrecioAutomatico = (peso: number, destino: string): number => {
  const preciosBase: Record<string, number> = {
    'La Paz': 20, 'Santa Cruz': 25, 'Cochabamba': 22, 'Oruro': 18,
    'Potosí': 20, 'Tarija': 23, 'Beni': 30, 'Pando': 35,
  };
  const base = preciosBase[destino] || 25;
  const adicionalPeso = peso > 1 ? (peso - 1) * 5 : 0;
  return base + adicionalPeso;
};

// Mappers para convertir entre formatos del frontend y backend
const toDTO = (data: CreateEncomiendaRequest) => ({
  remitente_nombre: data.remitente_nombre,
  remitente_telefono: data.remitente_telefono,
  remitente_direccion: data.remitente_direccion || '',
  destinatario_nombre: data.destinatario_nombre,
  destinatario_telefono: data.destinatario_telefono,
  destino_ciudad: data.destino_ciudad,
  destino_direccion: data.destino_direccion,
  descripcion: data.descripcion,
  peso: data.peso,
  // ✅ CORREGIDO: Incluir precio (calculado automáticamente si no viene)
  precio: data.precio || calcularPrecioAutomatico(data.peso, data.destino_ciudad),
  notas: data.notas || '',
  metodo_pago: data.metodo_pago,
});

const fromDTO = (data: BackendEncomienda): Encomienda => ({
  id: data.id,
  codigo_seguimiento: data.codigo_seguimiento,
  remitente_nombre: data.remitente_nombre,
  remitente_telefono: data.remitente_telefono,
  remitente_direccion: data.remitente_direccion,
  destinatario_nombre: data.destinatario_nombre,
  destinatario_telefono: data.destinatario_telefono,
  destino_ciudad: data.destino_ciudad,
  destino_direccion: data.destino_direccion,
  descripcion: data.descripcion,
  peso: typeof data.peso === 'string' ? parseFloat(data.peso) : data.peso,
  precio: typeof data.precio === 'string' ? parseFloat(data.precio) : data.precio,
  estado: data.estado,
  fecha_creacion: data.fecha_creacion,
  fecha_entrega_estimada: data.fecha_entrega_estimada,
  fecha_entrega_real: data.fecha_entrega_real,
  conductor_asignado: data.conductor_asignado,
  conductor_nombre: data.conductor_nombre,
  notas: data.notas,
  creado_por: data.creado_por,
  
  // NUEVOS CAMPOS PARA PAGOS
  pago_info: data.pago_info,
  estado_pago: data.estado_pago,
  metodo_pago: data.metodo_pago,
  seguimientos: data.seguimientos || [],
  // ✅ NUEVOS CAMPOS DEL BACKEND ACTUALIZADO
  pago: data.pago,
  puede_ser_asignada: data.puede_ser_asignada,
  puede_ser_entregada: data.puede_ser_entregada,
  creado_por_nombre: data.creado_por_nombre,
  conductor_info: data.conductor_info,
  pago_detalle: data.pago_detalle,
});

export const encomiendaService = {
  // Listar encomiendas con filtros y paginación (para admin)
  async list(filters?: EncomiendaFilters): Promise<ApiResponse<PaginatedResponse<Encomienda>>> {
    const params = new URLSearchParams();
    
    if (filters?.estado) params.append('estado', filters.estado);
    if (filters?.destino_ciudad) params.append('destino_ciudad', filters.destino_ciudad);
    if (filters?.fecha_desde) params.append('fecha_desde', filters.fecha_desde);
    if (filters?.fecha_hasta) params.append('fecha_hasta', filters.fecha_hasta);
    if (filters?.conductor_asignado) params.append('conductor_asignado', filters.conductor_asignado.toString());
    if (filters?.codigo_seguimiento) params.append('codigo_seguimiento', filters.codigo_seguimiento);
    
    const query = params.toString();
    const url = `/api/encomiendas/${query ? `?${query}` : ''}`;
    
    try {
      const response = await apiRequest<PaginatedResponse<BackendEncomienda>>(url);
      
      if (response.success && response.data) {
        const data = response.data;
        // CORREGIDO: Manejar diferentes formatos de respuesta
        const results = Array.isArray(data.results) ? data.results : 
                       Array.isArray(data) ? data : [];
        const count = data.count || results.length;
        
        return {
          success: true,
          data: {
            count: count,
            next: data.next,
            previous: data.previous,
            results: results.map(fromDTO),
          },
        };
      }
      
      return response as ApiResponse<PaginatedResponse<Encomienda>>;
    } catch (error) {
      return {
        success: false,
        error: 'Error de conexión al cargar encomiendas'
      };
    }
  },

  // Obtener encomienda por ID
  async getById(id: number): Promise<ApiResponse<Encomienda>> {
    try {
      const response = await apiRequest<BackendEncomienda>(`/api/encomiendas/${id}/`);
      
      if (response.success && response.data) {
        return { success: true, data: fromDTO(response.data) };
      }
      
      return response as ApiResponse<Encomienda>;
    } catch (error) {
      return {
        success: false,
        error: 'Error de conexión al obtener encomienda'
      };
    }
  },

  // Obtener encomienda por código de seguimiento
  async getByTrackingCode(codigo: string): Promise<ApiResponse<Encomienda>> {
    try {
      const response = await apiRequest<BackendEncomienda>(`/api/encomiendas/seguimiento/${codigo}/`);
      
      if (response.success && response.data) {
        return { success: true, data: fromDTO(response.data) };
      }
      
      return response as ApiResponse<Encomienda>;
    } catch (error) {
      return {
        success: false,
        error: 'Error de conexión al buscar encomienda'
      };
    }
  },

  // ✅ CORREGIDO: Crear nueva encomienda con mejor manejo de errores
  async create(data: CreateEncomiendaRequest): Promise<ApiResponse<Encomienda>> {
    try {
      const dto = toDTO(data);
      
      console.log('🔍 DEBUG - DTO a enviar:', dto);
      console.log('🔍 DEBUG - URL:', '/api/encomiendas/');

      const response = await apiRequest<BackendEncomienda>('/api/encomiendas/', {
        method: 'POST',
        body: JSON.stringify(dto),
      });

      console.log('🔍 DEBUG - Response status:', response);
      
      if (response.success && response.data) {
        return { 
          success: true, 
          data: fromDTO(response.data),
          message: 'Encomienda creada exitosamente'
        };
      } else {
        // ✅ MEJOR MANEJO DE ERRORES DETALLADO
        console.error('🔍 DEBUG - Error response:', response);
        const errorData = response as any;
        const errorMessage = errorData.error || 
                           (response.data && (response.data.error || response.data.detail)) ||
                           'Error desconocido al crear encomienda';
        
        return { 
          success: false, 
          error: errorMessage
        };
      }
    } catch (error: any) {
      console.error('🔍 DEBUG - Catch error:', error);
      return {
        success: false,
        error: error.message || 'Error de conexión al crear encomienda'
      };
    }
  },

  // Actualizar encomienda
  async update(id: number, data: Partial<Encomienda>): Promise<ApiResponse<Encomienda>> {
    try {
      const response = await apiRequest<BackendEncomienda>(`/api/encomiendas/${id}/`, {
        method: 'PATCH',
        body: JSON.stringify(data),
      });
      
      if (response.success && response.data) {
        return { success: true, data: fromDTO(response.data) };
      }
      
      return response as ApiResponse<Encomienda>;
    } catch (error) {
      return {
        success: false,
        error: 'Error de conexión al actualizar encomienda'
      };
    }
  },

  // Eliminar encomienda
  async remove(id: number): Promise<ApiResponse<void>> {
    try {
      const response = await apiRequest(`/api/encomiendas/${id}/`, { 
        method: 'DELETE' 
      });
      
      if (response.success) {
        return { success: true };
      }
      
      return response as ApiResponse<void>;
    } catch (error) {
      return {
        success: false,
        error: 'Error de conexión al eliminar encomienda'
      };
    }
  },

  // Obtener estadísticas de encomiendas
  async getStats(): Promise<ApiResponse<EncomiendaStats>> {
    try {
      const response = await apiRequest<EncomiendaStats>('/api/encomiendas/estadisticas/');
      if (response.success && response.data) {
        const stats = response.data;
        return {
          success: true,
          data: {
            total: stats.total ?? 0,
            pendientes: stats.pendientes ?? 0,
            en_ruta: stats.en_ruta ?? 0,
            entregados: stats.entregados ?? 0,
            cancelados: stats.cancelados ?? 0,
            ingresos_totales: stats.ingresos_totales ?? 0,
          },
        };
      }
      return response as ApiResponse<EncomiendaStats>;
    } catch (error) {
      return {
        success: false,
        error: 'Error de conexión al cargar estadísticas'
      };
    }
  },

  // Obtener mis encomiendas (para usuarios normales)
  async getMyEncomiendas(): Promise<ApiResponse<Encomienda[]>> {
    try {
      const response = await apiRequest<BackendEncomienda[] | { results: BackendEncomienda[] }>('/api/encomiendas/mis_encomiendas/');
      if (response.success && response.data) {
        const data = response.data;
        let lista: BackendEncomienda[] = [];
        
        if (Array.isArray(data)) {
          lista = data;
        } else if (data && 'results' in data && Array.isArray(data.results)) {
          lista = data.results;
        }
        
        return { 
          success: true, 
          data: lista.map(fromDTO) 
        };
      }
      
      if (response.success) {
        return { success: true, data: [] };
      }
      
      return response as ApiResponse<Encomienda[]>;
    } catch (error) {
      return {
        success: false,
        error: 'Error de conexión al cargar mis encomiendas'
      };
    }
  },

  // Obtener encomiendas asignadas al conductor autenticado
  async getAssignedEncomiendas(): Promise<ApiResponse<Encomienda[]>> {
    try {
      const response = await apiRequest<BackendEncomienda[] | { results: BackendEncomienda[] }>('/api/encomiendas/asignadas/');
      if (response.success && response.data) {
        const data = response.data;
        const lista = Array.isArray(data) ? data : 
                     (data && 'results' in data ? data.results : []);
        return { success: true, data: lista.map(fromDTO) };
      }
      return response as ApiResponse<Encomienda[]>;
    } catch (error) {
      return {
        success: false,
        error: 'Error de conexión al cargar encomiendas asignadas'
      };
    }
  },

  // Asignar conductor a encomienda
  async asignarConductor(encomiendaId: number, conductorId: number): Promise<ApiResponse<Encomienda>> {
    try {
      const response = await apiRequest<BackendEncomienda>(`/api/encomiendas/${encomiendaId}/asignar_conductor/`, {
        method: 'POST',
        body: JSON.stringify({ conductor_id: conductorId }),
      });
      if (response.success && response.data) {
        return { success: true, data: fromDTO(response.data) };
      }
      return response as ApiResponse<Encomienda>;
    } catch (error) {
      return {
        success: false,
        error: 'Error de conexión al asignar conductor'
      };
    }
  },

  // Actualizar estado de entrega
  async actualizarEstadoEntrega(encomiendaId: number, estado: string, notas?: string): Promise<ApiResponse<Encomienda>> {
    try {
      const response = await apiRequest<BackendEncomienda>(`/api/encomiendas/${encomiendaId}/actualizar_estado/`, {
        method: 'POST',
        body: JSON.stringify({ 
          estado, 
          notas,
          fecha_entrega_real: estado === 'entregado' ? new Date().toISOString() : undefined
        }),
      });
      if (response.success && response.data) {
        return { success: true, data: fromDTO(response.data) };
      }
      return response as ApiResponse<Encomienda>;
    } catch (error) {
      return {
        success: false,
        error: 'Error de conexión al actualizar estado'
      };
    }
  },

  // NUEVOS MÉTODOS PARA PAGOS

  // Crear pago en Stripe
  async crearPagoStripe(encomiendaId: number): Promise<ApiResponse<StripePaymentIntent>> {
    try {
      const response = await apiRequest<StripePaymentIntent>(`/api/encomiendas/${encomiendaId}/crear_pago_stripe/`, {
        method: 'POST',
      });
      
      if (response.success && response.data) {
        return {
          success: true,
          data: response.data
        };
      }
      
      return response as ApiResponse<StripePaymentIntent>;
    } catch (error) {
      return {
        success: false,
        error: 'Error de conexión al crear pago Stripe'
      };
    }
  },

  // Confirmar pago de Stripe
  async confirmarPago(encomiendaId: number, paymentIntentId: string): Promise<ApiResponse<Encomienda>> {
    try {
      interface ConfirmPagoResponse {
        encomienda?: BackendEncomienda;
        success?: boolean;
        message?: string;
      }

      const response = await apiRequest<ConfirmPagoResponse>(`/api/encomiendas/${encomiendaId}/confirmar_pago/`, {
        method: 'POST',
        body: JSON.stringify({ payment_intent_id: paymentIntentId }),
      });
      
      if (response.success && response.data) {
        const responseData = response.data;
        // ✅ CORREGIDO: Manejar diferentes formatos de respuesta
        if (responseData.encomienda) {
          return { success: true, data: fromDTO(responseData.encomienda) };
        } else if (responseData.success) {
          // Si no viene la encomienda, obtenerla de nuevo
          return this.getById(encomiendaId);
        }
      }
      
      return response as ApiResponse<Encomienda>;
    } catch (error) {
      return {
        success: false,
        error: 'Error de conexión al confirmar pago'
      };
    }
  },

  // Marcar pago en efectivo como completado (solo administradores)
  async marcarPagoEfectivo(encomiendaId: number): Promise<ApiResponse<Encomienda>> {
    try {
      interface PagoEfectivoResponse {
        encomienda?: BackendEncomienda;
        success?: boolean;
        message?: string;
      }

      const response = await apiRequest<PagoEfectivoResponse>(`/api/encomiendas/${encomiendaId}/marcar_pago_efectivo/`, {
        method: 'POST',
      });
      
      if (response.success && response.data) {
        const responseData = response.data;
        if (responseData.encomienda) {
          return { success: true, data: fromDTO(responseData.encomienda) };
        } else if (responseData.success) {
          return this.getById(encomiendaId);
        }
      }
      
      return response as ApiResponse<Encomienda>;
    } catch (error) {
      return {
        success: false,
        error: 'Error de conexión al marcar pago en efectivo'
      };
    }
  },

  // Generar reporte de encomiendas
  async generarReporte(filters?: EncomiendaFilters): Promise<ApiResponse<{ url: string }>> {
    try {
      const params = new URLSearchParams();
      
      if (filters?.estado) params.append('estado', filters.estado);
      if (filters?.destino_ciudad) params.append('destino_ciudad', filters.destino_ciudad);
      if (filters?.fecha_desde) params.append('fecha_desde', filters.fecha_desde);
      if (filters?.fecha_hasta) params.append('fecha_hasta', filters.fecha_hasta);
      
      const query = params.toString();
      const url = `/api/encomiendas/generar_reporte/${query ? `?${query}` : ''}`;
      const response = await apiRequest<{ url: string }>(url);
      
      return response as ApiResponse<{ url: string }>;
    } catch (error) {
      return {
        success: false,
        error: 'Error de conexión al generar reporte'
      };
    }
  },

  // CORREGIDO: Alias para 'delete' para mantener compatibilidad
  async delete(id: number): Promise<ApiResponse<void>> {
    return this.remove(id);
  },

  // ✅ NUEVO: Función para calcular precio (para usar en el frontend)
  calcularPrecio(peso: number, destino: string): number {
    return calcularPrecioAutomatico(peso, destino);
  }
};

File: mobile/lib/services/encomienda_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:mobile/services/auth_service.dart';
import 'package:mobile/services/logger_service.dart'; 
import '../models/encomienda_model.dart';
import '../models/crear_encomienda_model.dart';
import '../utils/ip_detection.dart';

class EncomiendaService {
  final AuthService _authService = AuthService();

  Future<Map<String, String>> _getHeaders() async {
    final token = await _authService.getToken();
    return {
      'Content-Type': 'application/json',
      if (token != null) 'Authorization': 'Bearer $token',
    };
  }

  // ✅ CORREGIDO: Crear encomienda
  Future<ApiResponse<Encomienda>> crearEncomienda(CrearEncomiendaRequest request) async {
    try {
      final baseUrl = await IPDetection.getBaseUrl();
      final headers = await _getHeaders();
      
      Logger.network('Creando encomienda: ${request.toJson()}', tag: 'ENCOMIENDA');
      
      final response = await http.post(
        Uri.parse('$baseUrl/api/encomiendas/'),
        headers: headers,
        body: json.encode(request.toJson()),
      );

      Logger.network('Response status: ${response.statusCode}', tag: 'ENCOMIENDA');
      Logger.network('Response body: ${response.body}', tag: 'ENCOMIENDA');

      if (response.statusCode == 201 || response.statusCode == 200) {
        final responseData = json.decode(response.body);
        final encomienda = Encomienda.fromJson(responseData);
        
        Logger.success('Encomienda creada exitosamente', tag: 'ENCOMIENDA');
        
        return ApiResponse<Encomienda>(
          success: true,
          data: encomienda,
          message: 'Encomienda creada exitosamente'
        );
      } else {
        final errorData = json.decode(response.body);
        final errorMessage = _parseError(errorData, response.statusCode);
        
        Logger.error('Error al crear encomienda: $errorMessage', tag: 'ENCOMIENDA');
        
        return ApiResponse<Encomienda>(
          success: false,
          error: errorMessage
        );
      }
    } catch (e) {
      Logger.error('Error en crearEncomienda', tag: 'ENCOMIENDA', error: e);
      return ApiResponse<Encomienda>(
        success: false,
        error: 'Error de conexión: $e'
      );
    }
  }

  // ✅ CORREGIDO: Obtener mis encomiendas
  Future<ApiResponse<List<Encomienda>>> getMisEncomiendas() async {
    try {
      final baseUrl = await IPDetection.getBaseUrl();
      final headers = await _getHeaders();
      
      Logger.network('Obteniendo mis encomiendas...', tag: 'ENCOMIENDA');
      
      final response = await http.get(
        Uri.parse('$baseUrl/api/encomiendas/mis_encomiendas/'),
        headers: headers,
      );

      Logger.network('Mis encomiendas response: ${response.statusCode}', tag: 'ENCOMIENDA');
      Logger.network('Mis encomiendas body: ${response.body}', tag: 'ENCOMIENDA');

      if (response.statusCode == 200) {
        final responseData = json.decode(response.body);
        
        // ✅ DEBUG DETALLADO - Ver estructura real
        Logger.debug('Estructura de respuesta: ${responseData.runtimeType}', tag: 'ENCOMIENDA');
        if (responseData is Map) {
          Logger.debug('Keys del mapa: ${responseData.keys}', tag: 'ENCOMIENDA');
        }
        
        List<dynamic> results = [];
        
        if (responseData is Map) {
          if (responseData.containsKey('data')) {
            results = responseData['data'] ?? [];
          } else if (responseData.containsKey('results')) {
            results = responseData['results'] ?? [];
          } else if (responseData.containsKey('encomiendas')) {
            results = responseData['encomiendas'] ?? [];
          } else {
            final possibleList = responseData.values.firstWhere(
              (value) => value is List,
              orElse: () => []
            );
            results = possibleList is List ? possibleList : [];
          }
        } else if (responseData is List) {
          results = responseData;
        }
        
        Logger.debug('Resultados encontrados: ${results.length}', tag: 'ENCOMIENDA');
        
        if (results.isNotEmpty) {
          Logger.debug('Primer elemento: ${results.first}', tag: 'ENCOMIENDA');
        }
        
        // ✅ MANEJO DE ERRORES EN PARSEO
        final encomiendas = <Encomienda>[];
        for (var i = 0; i < results.length; i++) {
          try {
            final encomienda = Encomienda.fromJson(results[i]);
            encomiendas.add(encomienda);
          } catch (e) {
            Logger.error(
              'Error parseando encomienda $i', 
              tag: 'ENCOMIENDA', 
              error: e,
            );
            Logger.debug('Datos problemáticos: ${results[i]}', tag: 'ENCOMIENDA');
          }
        }
        
        Logger.success('Encomiendas cargadas: ${encomiendas.length}', tag: 'ENCOMIENDA');
        
        return ApiResponse<List<Encomienda>>(
          success: true,
          data: encomiendas,
          message: 'Encomiendas cargadas exitosamente'
        );
      } else if (response.statusCode == 404) {
        Logger.info('No hay encomiendas registradas', tag: 'ENCOMIENDA');
        return ApiResponse<List<Encomienda>>(
          success: true,
          data: [],
          message: 'No tienes encomiendas registradas'
        );
      } else {
        final errorData = json.decode(response.body);
        final error = _parseError(errorData, response.statusCode);
        
        Logger.error('Error del servidor: $error', tag: 'ENCOMIENDA');
        
        return ApiResponse<List<Encomienda>>(
          success: false,
          error: error
        );
      }
    } catch (e) {
      Logger.error('Error en getMisEncomiendas', tag: 'ENCOMIENDA', error: e);
      return ApiResponse<List<Encomienda>>(
        success: false,
        error: 'Error de conexión: $e'
      );
    }
  }

  // ✅ Método para parsear errores
  String _parseError(dynamic errorData, int statusCode) {
    if (errorData is Map) {
      return errorData['error'] ?? 
             errorData['detail'] ?? 
             errorData['message'] ??
             errorData.toString();
    } else if (errorData is String) {
      return errorData;
    } else {
      return 'Error del servidor: $statusCode';
    }
  }

  Future<ApiResponse<Encomienda>> getSeguimiento(String codigo) async {
    try {
      final baseUrl = await IPDetection.getBaseUrl();
      
      Logger.network('Buscando seguimiento: $codigo', tag: 'SEGUIMIENTO');
      
      final response = await http.get(
        Uri.parse('$baseUrl/api/encomiendas/seguimiento/$codigo/'),
      );

      Logger.network('Seguimiento response: ${response.statusCode}', tag: 'SEGUIMIENTO');
      Logger.network('Seguimiento body: ${response.body}', tag: 'SEGUIMIENTO');

      if (response.statusCode == 200) {
        final responseData = json.decode(response.body);
        final encomienda = Encomienda.fromJson(responseData);
        
        Logger.success('Encomienda encontrada', tag: 'SEGUIMIENTO');
        
        return ApiResponse<Encomienda>(
          success: true,
          data: encomienda,
          message: 'Encomienda encontrada'
        );
      } else if (response.statusCode == 404) {
        Logger.warning('Encomienda no encontrada: $codigo', tag: 'SEGUIMIENTO');
        return ApiResponse<Encomienda>(
          success: false,
          error: 'Encomienda no encontrada'
        );
      } else {
        final errorData = json.decode(response.body);
        final error = _parseError(errorData, response.statusCode);
        
        Logger.error('Error en seguimiento: $error', tag: 'SEGUIMIENTO');
        
        return ApiResponse<Encomienda>(
          success: false,
          error: error
        );
      }
    } catch (e) {
      Logger.error('Error en getSeguimiento', tag: 'SEGUIMIENTO', error: e);
      return ApiResponse<Encomienda>(
        success: false,
        error: 'Error de conexión: $e'
      );
    }
  }

  Future<ApiResponse<Map<String, dynamic>>> getEstadisticas() async {
    try {
      final baseUrl = await IPDetection.getBaseUrl();
      final headers = await _getHeaders();
      
      Logger.network('Obteniendo estadísticas...', tag: 'ESTADISTICAS');
      
      final response = await http.get(
        Uri.parse('$baseUrl/api/encomiendas/estadisticas/'),
        headers: headers,
      );

      if (response.statusCode == 200) {
        final responseData = json.decode(response.body);
        
        Logger.success('Estadísticas cargadas', tag: 'ESTADISTICAS');
        
        return ApiResponse<Map<String, dynamic>>(
          success: true,
          data: responseData,
          message: 'Estadísticas cargadas'
        );
      } else {
        final errorData = json.decode(response.body);
        final error = _parseError(errorData, response.statusCode);
        
        Logger.error('Error al cargar estadísticas: $error', tag: 'ESTADISTICAS');
        
        return ApiResponse<Map<String, dynamic>>(
          success: false,
          error: error
        );
      }
    } catch (e) {
      Logger.error('Error en getEstadisticas', tag: 'ESTADISTICAS', error: e);
      return ApiResponse<Map<String, dynamic>>(
        success: false,
        error: 'Error de conexión: $e'
      );
    }
  }

  // ✅ NUEVO: Actualizar estado de encomienda
  Future<ApiResponse<Encomienda>> actualizarEstado(int encomiendaId, String estado, {String? notas, String? ubicacion}) async {
    try {
      final baseUrl = await IPDetection.getBaseUrl();
      final headers = await _getHeaders();
      
      final body = {
        'estado': estado,
        if (notas != null) 'notas': notas,
        if (ubicacion != null) 'ubicacion': ubicacion,
      };
      
      Logger.network('Actualizando estado de encomienda $encomiendaId a $estado', tag: 'ESTADO');
      
      final response = await http.post(
        Uri.parse('$baseUrl/api/encomiendas/$encomiendaId/actualizar_estado/'),
        headers: headers,
        body: json.encode(body),
      );

      if (response.statusCode == 200) {
        final responseData = json.decode(response.body);
        final encomienda = Encomienda.fromJson(responseData);
        
        Logger.success('Estado actualizado exitosamente', tag: 'ESTADO');
        
        return ApiResponse<Encomienda>(
          success: true,
          data: encomienda,
          message: 'Estado actualizado exitosamente'
        );
      } else {
        final errorData = json.decode(response.body);
        final error = errorData['error'] ?? 'Error al actualizar estado';
        
        Logger.error('Error al actualizar estado: $error', tag: 'ESTADO');
        
        return ApiResponse<Encomienda>(
          success: false,
          error: error
        );
      }
    } catch (e) {
      Logger.error('Error en actualizarEstado', tag: 'ESTADO', error: e);
      return ApiResponse<Encomienda>(
        success: false,
        error: 'Error de conexión: $e'
      );
    }
  }

  // ✅ MÉTODOS DE PAGO
  Future<ApiResponse<Map<String, dynamic>>> crearPagoStripe(int encomiendaId) async {
    try {
      final baseUrl = await IPDetection.getBaseUrl();
      final headers = await _getHeaders();
      
      Logger.network('Creando pago Stripe para encomienda: $encomiendaId', tag: 'PAGO');
      
      final response = await http.post(
        Uri.parse('$baseUrl/api/encomiendas/$encomiendaId/crear_pago_stripe/'),
        headers: headers,
      );

      if (response.statusCode == 200) {
        final responseData = json.decode(response.body);
        
        Logger.success('Pago Stripe creado exitosamente', tag: 'PAGO');
        
        return ApiResponse<Map<String, dynamic>>(
          success: true,
          data: responseData,
          message: 'Pago creado exitosamente'
        );
      } else {
        final errorData = json.decode(response.body);
        final error = errorData['error'] ?? 'Error al crear pago';
        
        Logger.error('Error al crear pago Stripe: $error', tag: 'PAGO');
        
        return ApiResponse<Map<String, dynamic>>(
          success: false,
          error: error
        );
      }
    } catch (e) {
      Logger.error('Error en crearPagoStripe', tag: 'PAGO', error: e);
      return ApiResponse<Map<String, dynamic>>(
        success: false,
        error: 'Error de conexión: $e'
      );
    }
  }

  Future<ApiResponse<Encomienda>> confirmarPago(int encomiendaId, String paymentIntentId) async {
    try {
      final baseUrl = await IPDetection.getBaseUrl();
      final headers = await _getHeaders();
      
      Logger.network('Confirmando pago para encomienda: $encomiendaId', tag: 'PAGO');
      
      final response = await http.post(
        Uri.parse('$baseUrl/api/encomiendas/$encomiendaId/confirmar_pago/'),
        headers: headers,
        body: json.encode({'payment_intent_id': paymentIntentId}),
      );

      if (response.statusCode == 200) {
        final responseData = json.decode(response.body);
        final encomienda = Encomienda.fromJson(responseData);
        
        Logger.success('Pago confirmado exitosamente', tag: 'PAGO');
        
        return ApiResponse<Encomienda>(
          success: true,
          data: encomienda,
          message: 'Pago confirmado exitosamente'
        );
      } else {
        final errorData = json.decode(response.body);
        final error = errorData['error'] ?? 'Error al confirmar pago';
        
        Logger.error('Error al confirmar pago: $error', tag: 'PAGO');
        
        return ApiResponse<Encomienda>(
          success: false,
          error: error
        );
      }
    } catch (e) {
      Logger.error('Error en confirmarPago', tag: 'PAGO', error: e);
      return ApiResponse<Encomienda>(
        success: false,
        error: 'Error de conexión: $e'
      );
    }
  }
   // ✅ NUEVOS MÉTODOS PARA PAGOS CON STRIPe
  Future<ApiResponse<Encomienda>> confirmarPagoStripe(int encomiendaId, String paymentIntentId) async {
    try {
      final baseUrl = await IPDetection.getBaseUrl();
      final headers = await _getHeaders();
      
      Logger.network('Confirmando pago Stripe para encomienda: $encomiendaId', tag: 'PAGO_STRIPE');
      
      final response = await http.post(
        Uri.parse('$baseUrl/api/encomiendas/$encomiendaId/confirmar_pago/'),
        headers: headers,
        body: json.encode({'payment_intent_id': paymentIntentId}),
      );

      Logger.network('Confirmar pago response: ${response.statusCode}', tag: 'PAGO_STRIPE');
      Logger.network('Confirmar pago body: ${response.body}', tag: 'PAGO_STRIPE');

      if (response.statusCode == 200) {
        final responseData = json.decode(response.body);
        final encomienda = Encomienda.fromJson(responseData);
        
        Logger.success('Pago Stripe confirmado exitosamente', tag: 'PAGO_STRIPE');
        
        return ApiResponse<Encomienda>(
          success: true,
          data: encomienda,
          message: 'Pago con tarjeta confirmado exitosamente'
        );
      } else {
        final errorData = json.decode(response.body);
        final errorMessage = _parseError(errorData, response.statusCode);
        
        Logger.error('Error al confirmar pago Stripe: $errorMessage', tag: 'PAGO_STRIPE');
        
        return ApiResponse<Encomienda>(
          success: false,
          error: errorMessage
        );
      }
    } catch (e) {
      Logger.error('Error en confirmarPagoStripe', tag: 'PAGO_STRIPE', error: e);
      return ApiResponse<Encomienda>(
        success: false,
        error: 'Error de conexión: $e'
      );
    }
  }


  Future<ApiResponse<Encomienda>> marcarPagoEfectivo(int encomiendaId) async {
    try {
      final baseUrl = await IPDetection.getBaseUrl();
      final headers = await _getHeaders();
      
      Logger.network('Marcando pago en efectivo para encomienda: $encomiendaId', tag: 'PAGO');
      
      final response = await http.post(
        Uri.parse('$baseUrl/api/encomiendas/$encomiendaId/marcar_pago_efectivo/'),
        headers: headers,
      );

      Logger.network('Pago efectivo response: ${response.statusCode}', tag: 'PAGO');
      Logger.network('Pago efectivo body: ${response.body}', tag: 'PAGO');

      if (response.statusCode == 200) {
        final responseData = json.decode(response.body);
        final encomienda = Encomienda.fromJson(responseData['data']);
        
        Logger.success('Pago en efectivo registrado exitosamente', tag: 'PAGO');
        
        return ApiResponse<Encomienda>(
          success: true,
          data: encomienda,
          message: responseData['message'] ?? 'Pago en efectivo registrado exitosamente'
        );
      } else {
        final errorData = json.decode(response.body);
        final errorMessage = _parseError(errorData, response.statusCode);
        
        Logger.error('Error en pago efectivo: $errorMessage', tag: 'PAGO');
        
        return ApiResponse<Encomienda>(
          success: false,
          error: errorMessage
        );
      }
    } catch (e) {
      Logger.error('Error en marcarPagoEfectivo', tag: 'PAGO', error: e);
      return ApiResponse<Encomienda>(
        success: false,
        error: 'Error de conexión: $e'
      );
    }
  }
}

class ApiResponse<T> {
  final bool success;
  final T? data;
  final String? error;
  final String? message;

  ApiResponse({
    required this.success,
    this.data,
    this.error,
    this.message,
  });
}